### DISCUSSION TITLE
How does one use Navit to navigate?

### DISCUSSION THREAD

**Discussion by dca00** [navit-gps/navit#1245] [Q&A]
Could not figure out your UI.
So, I am somewhere in Canada, and I have an OSM map that covers my province and some of the adjacent USA.
I enter my city. The search result only shows me the matching US locations but not my actual city in Canada.
Somehow I fiddled with UI elements and got into a #, street, city mode. I enter my street # and street, and nothing is found.
On the map, my location is currently in that very street, in front of that very #.
Any tips on what I am doing wrong?

**Comment by mvglasow** [comment]
Workflow should be:
* **Actions > Town**
* Tap/click the country flag, then enter the country for your destination and tap/click the right one from the list. (Search works by name as well as by ISO country code.)
* Then, enter the city, and tap/click the correct one from the list to select it.
* **Streets**
* Enter the street name, tap/click the correct street from the list.
* **House numbers**
* Enter the house number, tap/click the correct one from the list.
* **Set as destination**

Let me guess: you probably did not select Canada, and Navit is pre-selecting the US as your country. The default country is taken from the locale settings.

You can change the default country in `navit.xml`. In the root element. `config`, make sure there’s an attribute named `language`. Format is something like `language="en_US"`: the first two letters (lowercase) specify the language, the last two (uppercase) specify the country, which is used as the default country for the search. Any combination of language and country will work (as long as there is a UI translation for the language), even exotic combinations such as `et_IN` (Estonian UI, search defaulting to India).

`language="en_CA"` should give you an English-language UI and set Canada as the default country for the address search. Or `language="fr_CA"` if you would like the UI in French.

**Reply by dca00** [reply]
What happens when I travel between the two countries? Do I have to keep changing the config every time I cross the border?

**Reply by dca00** [reply]
I selected Canada by clicking on the flag. It worked once then reset to the US.

**Reply by dca00** [reply]
Added the language attribute. Now the country is Canada. I try to enter a street address but it is not found.

**Comment by mvglasow** [comment]
Address search always starts with the default country; it doesn’t memorize the last one you selected. You can always select the country manually, but you’ll have to do it each time you search for an address outside the default country. Usually, it’s most convenient to set it to the country you live in (or where the majority of your travel destinations are).

If you travel to the same destinations frequently, you can also bookmark them. That way, you don’t have to enter the address manually every time.

At what step did address search fail? The internal GUI does not parse strings like “6125 Rue St Jacques, Montréal” – you’d first have to find Montréal, then Rue Saint-Jacques (with the correct spelling – Navit is not very good at dealing with spelling variations, but searching for substrings like “Jacques” will work), and finally the house number. (Searching for the house number will work only if the street has house number information in OSM – otherwise you can chose the street as the destination and Navit will guide you to that street.)

Which of these was not found?

There can be some confusion if there are multiple towns/cities with the same name. Where possible, Navit tries to add information such as post codes or administrative areas in which the town is located. However, this information is added when the map is generated, and the rules for that are hardcoded in maptool and depend on the country. Not sure about Canada, but I would expect at least the province to show up. If you end up selecting the wrong town, you’ll unwittingly be searching for a street in a tiny little village that happens to have the same name as a big city hundreds of km away.

**Reply by dca00** [reply]
Now I am beginning to understand why it feels so confusing and counterintuitive. What you describe is the opposite of how GPS navigation usually works. Are you sure that is the way navit should operate? Should it not mimic rather hardware devices or google maps? Because if someone is on the road and going to 123 Main St., Dumbton, Scratchshire, that is exactly how they expect to enter the address. What you describe is foreign to mainstream mentality of travelers. When I first saw navit I was surprised that it is not the top navigation software out there, but now I understand: with the reverse address entry it will never gain acceptance.
Imagine that I am a courier delivery driver. I have neither time nor desire to unlearn my lifetime habit and learn the opposite of it, for the sake of a new piece of software, no matter how advanced or perfect in all other respects. I have to drop my parcel off, on somebody's porch, hop back into the car, enter the next address the way I see it on the label, and get going. Makes sense?

**Reply by mvglasow** [reply]
You may well be right. The code base is 20 years old – Navit started before smartphones, Google Maps and the like were a thing, and this way of entering the address was the mainstream for navigation software back then.

It would certainly be possible to give the UX an overhaul and introduce free text address search (the Android UI might have something close to that, but I don’t use it much), but somebody would need to step forward and put time and effort into it.

That said, the “Google way” is not flawless either – it just has different flaws. Because it lacks a way to tell the system which part of the address is the town and which is the street, and because it is built to include similar results in the search (to compensate for misspellings or just legitimate spelling variations like St vs. Saint), you may end up with gems like “You entered York St., Majorcity. Did you mean York Street, Tinyville, Majorcity County?” even though the address you entered exists. So there’s something to be said for clearly designated fields with a clear meaning.

**Reply by dca00** [reply]
All of that already exists! If there were clearly defined series of entry fields: house#, street, city, with cursor key navigation among them, i.e. left arrow to go back and right arrow to go right, then the problem would have been solved!
Enter 1, 2, 3, hit Right, enter Main st, hit right, enter Dumbton. Uncertainty ends right there.
Ideally, the nearest match should be offered as a drop-down selection. Just like the current country should be easily discoverable by Navit. Why it is not coded that way is beyond me.
There are 100s of 1000s of Main st, but as soon as you narrow it down to Dumbton, you get 1 province/state, 2 tops.
The counterintuitive UI is likely the only thing that stands between Navit and its wide adoption. That and the installer.
Device access also has to be figured out because under Linux it cannot read from /dev/tty unless sudoed. But this is not a showstopper, not especially on devices that run Navit exclusively. Most of us do not give a damn if anyone in Linux distros frowns upon us running programs as root.
